cpp/include/cuvs/neighbors/cagra.hpp

```diff
@@ -385,35 +387,65 @@ struct index : cuvs::neighbors::index {
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<float, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::host_matrix_view<const float, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<float, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::device_matrix_view<const int8_t, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<int8_t, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::host_matrix_view<const int8_t, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<int8_t, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::device_matrix_view<const uint8_t, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<uint8_t, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::host_matrix_view<const uint8_t, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
```

Comment:
This is awkward- a user is going to have to set std::nullopt for at least one of nn-descent_params or ivf_pq params. I think we should make these build_params for cagra. Actually, I think what we should do is take away the `graph_build_algo` option from `build_params` and add a `graph_build_params` option that accepts the `cuvs::neighbors::build_params` base class, documents it approprirately with what's allowed, and then dispatches accordingly. Also remember to raise an exception if anything other than `ivf_pq` or `nn_descent` is used. This allows us to add more graph build algos in the future without too much additional effort.

Reply:
On second though- since you need refine_ratio on the `ivf_pq_build_params`, why not have a new type `cagra::graph_build_params` which can extend `cuvs::neighbors::build_params`? Then you can have a `cagra::graph_build_params::ivf_pq` which extends `ivf_pq::build_params` and `cagra::graph_build_params` and adds the `refine ratio`. Then you can have `cagra__graph_build_params::nn_descent` which does the same for nn-descent. This way you can have `cagra::build_params.graph_build_params` which accepts anything that extends `cagra::graph_build_params` and dispatches the build algo accordingly? This would get all this additional stuff out of `cagra::build` and keep the API clean and easy to follow, while minimizing duplication.

Reply:
I think your last idea looks great. It's clean, explicit, and tells the user right away what graph_build_params variants are supported.